File: java/872.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        if(root1 != null && root2 != null){
            ArrayList<Integer> leaf1 = new ArrayList<Integer>();
            ArrayList<Integer> leaf2 = new ArrayList<Integer>();
            dfs(root1,leaf1);
            dfs(root2,leaf2);
            return leaf1.equals(leaf2);
        }
        return false;
    }
    
    public void dfs(TreeNode root, ArrayList<Integer> leaf){
        if(root == null){
            return ;
        }
        if(root.left == null && root.right == null){
            leaf.add(root.val);
        }
        dfs(root.left,leaf);
        dfs(root.right,leaf);
            
    }
}

File: go/872.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func leafSimilar(root1 *TreeNode, root2 *TreeNode) bool {
	leaf1 := make([]int, 0)
	leaf2 := make([]int, 0)
	dfs(root1, &leaf1)
	dfs(root2, &leaf2)
	if len(leaf1) != len(leaf2) {
		return false
	}
	for i := 0; i < len(leaf1); i++ {
		if leaf1[i] != leaf2[i] {
			return false
		}
	}
	return true
}
func dfs(node *TreeNode, leaf *([]int)) {
	if node == nil {
		return
	}
	if node.Left == nil && node.Right == nil {
		*leaf = append(*leaf, node.Val)
	}
	dfs(node.Left, leaf)
	dfs(node.Right, leaf)
}